File: verdant_erp_backend/app/schemas/hvac/intelligence.py
from datetime import datetime
from pydantic import BaseModel


# =========================================================
# VISUAL INTELLIGENCE
# =========================================================

class GeoFailureDTO(BaseModel):
    location_id: str | None
    failures: int


class FacilityHealthDTO(BaseModel):
    location_id: str | None
    health_score: int
    failures: int
    status: str


class GeoHeatPointDTO(BaseModel):
    location_id: str | None
    intensity: int
    lat: float | None = None
    lng: float | None = None


class FailureClusterDTO(BaseModel):
    customer_id: str | None
    location_id: str | None
    cluster_strength: int
    cluster_type: str


class VisualIntelligenceSnapshotDTO(BaseModel):
    facility_health: list[FacilityHealthDTO]
    geo_heatmap: list[GeoHeatPointDTO]
    clusters: list[FailureClusterDTO]
    geo_failures: list[GeoFailureDTO]


# =========================================================
# SERVICE INTELLIGENCE
# =========================================================

class TechnicianPerformanceDTO(BaseModel):
    technician_id: str
    name: str
    completed_orders: int
    avg_resolution_time_hours: float


class SlaDashboardDTO(BaseModel):
    total_completed_orders: int
    on_time: int
    late: int
    sla_percentage: float


class CustomerSatisfactionDTO(BaseModel):
    customer_id: str | None
    total_orders: int
    reopened_orders: int
    avg_resolution_hours: float
    satisfaction_score: float


class TimeToRepairDTO(BaseModel):
    status: str
    avg_repair_time_hours: float


class ServiceIntelligenceSnapshotDTO(BaseModel):
    technician_performance: list[TechnicianPerformanceDTO]
    sla: SlaDashboardDTO
    customer_satisfaction: list[CustomerSatisfactionDTO]
    time_to_repair: list[TimeToRepairDTO]
    generated_at: datetime


# =========================================================
# INFRASTRUCTURE INTELLIGENCE
# =========================================================

class FailureByCustomerDTO(BaseModel):
    customer_id: str | None
    failures: int


class ProblematicEquipmentDTO(BaseModel):
    equipment_id: str
    asset_tag: str
    failures: int


class MaintenanceCostDTO(BaseModel):
    equipment_id: str
    asset_tag: str
    estimated_cost: float


class RiskScoringDTO(BaseModel):
    equipment_id: str
    asset_tag: str
    risk: str
    failures: int


class InfrastructureSnapshotDTO(BaseModel):
    failures_by_customer: list[FailureByCustomerDTO]
    problematic_equipment: list[ProblematicEquipmentDTO]
    maintenance_costs: list[MaintenanceCostDTO]
    risk_scoring: list[RiskScoringDTO]


# =========================================================
# CORRELATION ENGINE
# =========================================================

class CustomerClusterDTO(BaseModel):
    customer_id: str | None
    total_failures: int
    risk_level: str


class CorrelationProblematicEquipmentDTO(BaseModel):
    equipment_id: str
    asset_tag: str
    customer_id: str | None
    failures: int


class InstallationRankingDTO(BaseModel):
    location_id: str | None
    failures: int
    risk: str


class TechnicianHeatmapDTO(BaseModel):
    technician_id: str
    name: str
    workload: int
    load_level: str


class CorrelationSnapshotDTO(BaseModel):
    customer_clusters: list[CustomerClusterDTO]
    problematic_equipment: list[
        CorrelationProblematicEquipmentDTO
    ]
    installation_ranking: list[
        InstallationRankingDTO
    ]
    technician_heatmap: list[
        TechnicianHeatmapDTO
    ]


# =========================================================
# MASTER SNAPSHOT
# =========================================================

class IntelligenceSnapshotDTO(BaseModel):
    infrastructure: InfrastructureSnapshotDTO
    service: ServiceIntelligenceSnapshotDTO

File: verdant-frontend/src/lib/types/intelligence.ts
export type GeoFailure = {
    location_id: string | null;
    failures: number;
};

export type FacilityHealth = {
    location_id: string | null;
    health_score: number;
    failures: number;
    status: string;
};

export type GeoHeatPoint = {
    location_id: string | null;
    intensity: number;
    lat: number | null;
    lng: number | null;
};

export type FailureCluster = {
    customer_id: string | null;
    location_id: string | null;
    cluster_strength: number;
    cluster_type: string;
};

export type VisualIntelligenceSnapshot = {
    facility_health: FacilityHealth[];
    geo_heatmap: GeoHeatPoint[];
    clusters: FailureCluster[];
    geo_failures: GeoFailure[];
};

export type TechnicianPerformance = {
    technician_id: string;
    name: string;
    completed_orders: number;
    avg_resolution_time_hours: number;
};

export type SlaDashboard = {
    total_completed_orders: number;
    on_time: number;
    late: number;
    sla_percentage: number;
};

export type CustomerSatisfaction = {
    customer_id: string | null;
    total_orders: number;
    reopened_orders: number;
    avg_resolution_hours: number;
    satisfaction_score: number;
};

export type TimeToRepair = {
    status: string;
    avg_repair_time_hours: number;
};

export type ServiceIntelligenceSnapshot = {
    technician_performance: TechnicianPerformance[];
    sla: SlaDashboard;
    customer_satisfaction: CustomerSatisfaction[];
    time_to_repair: TimeToRepair[];
    generated_at: string;
};

export type FailureByCustomer = {
    customer_id: string | null;
    failures: number;
};

export type ProblematicEquipment = {
    equipment_id: string;
    asset_tag: string;
    failures: number;
};

export type MaintenanceCost = {
    equipment_id: string;
    asset_tag: string;
    estimated_cost: number;
};

export type RiskScoring = {
    equipment_id: string;
    asset_tag: string;
    risk: string;
    failures: number;
};

export type InfrastructureSnapshot = {
    failures_by_customer: FailureByCustomer[];
    problematic_equipment: ProblematicEquipment[];
    maintenance_costs: MaintenanceCost[];
    risk_scoring: RiskScoring[];
};

export type CustomerCluster = {
    customer_id: string | null;
    total_failures: number;
    risk_level: string;
};

export type CorrelationProblematicEquipment = {
    equipment_id: string;
    asset_tag: string;
    customer_id: string | null;
    failures: number;
};

export type InstallationRanking = {
    location_id: string | null;
    failures: number;
    risk: string;
};

export type TechnicianHeatmap = {
    technician_id: string;
    name: string;
    workload: number;
    load_level: string;
};

export type CorrelationSnapshot = {
    customer_clusters: CustomerCluster[];
    problematic_equipment:
        CorrelationProblematicEquipment[];
    installation_ranking:
        InstallationRanking[];
    technician_heatmap:
        TechnicianHeatmap[];
};

export type IntelligenceSnapshot = {
    infrastructure: InfrastructureSnapshot;
    service: ServiceIntelligenceSnapshot;
};